### PR TITLE
Request chain refactor

### PR DESCRIPTION
**Pull request by 22a**
"refactor"

per https://github.com/tom-and-the-toothfairies/pathways/issues/109, split the chained requests into individual response handlers, should make it easier to reason about UI logic

### CODE REVIEW

**Comment by c-brenn** [comment]
this is much nicer 🥇 

Looking at all the `handle_` functions, I notice they have the same structure:

```javascript
if (response.ok) {
  ...
} else {
  ...
  displayError(...)
}
```

Do you think it would be worth breaking out a more generic higher-order function?

```javascript
handleResponse = (response, callback) => {
  if (response.ok) {
    const data = await ...;
    callback(data);
  } else {
     const {error} = await ...;
     displayError(error);
  }
};
```

File: panacea/web/static/js/app.js [hunk 1]
```diff
@@ -25,64 +25,80 @@ async function submitFile() {
...
+        method: 'POST',
+        body: JSON.stringify({drugs: uris}),
+        credentials: 'same-origin',
+        headers: defaultHeaders
+      });
+
+      handleDdisResponse(ddisResponse);
+
+    } else {
+      console.log("by definition ddis require more than one drug")
```
Comment (houli) [inline panacea/web/static/js/app.js]: semicolon

Comment (22a) [reply panacea/web/static/js/app.js]: ty x

File: panacea/web/static/js/app.js [hunk 2]
```diff
@@ -25,64 +25,80 @@ async function submitFile() {
...
+  if (response.ok) {
+    const payload = await response.json();
+    handle(payload);
+  } else {
+    const {error} = await response.json();
+    displayError(error);
+  }
+}
+
+async function handleDrugsResponse(drugsResponse) {
```
Comment (houli) [inline panacea/web/static/js/app.js]: Doesn't need to be an async function

File: panacea/web/static/js/app.js [hunk 3]
```diff
@@ -25,64 +25,80 @@ async function submitFile() {
...
+
+      handleUrisResponse(urisResponse);
+
+    } else {
+      displayError({title: "Pathway error", detail: "No drugs found"});
+    }
+  });
+}
+
+async function handleUrisResponse(urisResponse) {
```
Comment (houli) [inline panacea/web/static/js/app.js]: Doesn't need to be an async function

File: panacea/web/static/js/app.js [hunk 4]
```diff
@@ -25,64 +25,80 @@ async function submitFile() {
...
+      handleDdisResponse(ddisResponse);
+
+    } else {
+      console.log("by definition ddis require more than one drug")
+    }
+  });
+}
+
+async function handleDdisResponse(ddisResponse) {
+  handleResponse(ddisResponse, async function (payload) {
```
Comment (houli) [inline panacea/web/static/js/app.js]: Neither of these have to be async functions

**Review by houli — CHANGES_REQUESTED** [review]

File: panacea/web/static/js/app.js (hunk 1 above)
Comment (22a) [reply panacea/web/static/js/app.js]: i should really use a linter 🌡

File: panacea/web/static/js/app.js (hunk 4 above)
Comment (22a) [reply panacea/web/static/js/app.js]: noted, will rectify

File: panacea/web/static/js/app.js [hunk 5]
```diff
@@ -25,64 +25,80 @@ async function submitFile() {
...
+    handle(payload);
+  } else {
+    const {error} = await response.json();
+    displayError(error);
+  }
+}
+
+async function handleDrugsResponse(drugsResponse) {
+  handleResponse(drugsResponse, async function (payload) {
+    const {drugs} = payload;
```
Comment (c-brenn) [inline panacea/web/static/js/app.js]: How about destructuring in the function definition ? Would remove the assignment. Same in the other handlers

**Review by c-brenn — CHANGES_REQUESTED** [review]

File: panacea/web/static/js/app.js (hunk 5 above)
Comment (22a) [reply panacea/web/static/js/app.js]: not all the destructuring is the same

Comment (22a) [reply panacea/web/static/js/app.js]: https://github.com/tom-and-the-toothfairies/pathways/pull/117/files#diff-af326ae966b05ea64b817eac7636661eR72 for example, i didn't know how best to generalise the destructuring

Comment (c-brenn) [reply panacea/web/static/js/app.js]: No need to generalise. I mean in the callback definition.

Comment (c-brenn) [reply panacea/web/static/js/app.js]: function(payload) becomes function({drugs})

Comment (22a) [reply panacea/web/static/js/app.js]: ahhhhh, i understand now, that makes sense

File: panacea/web/static/js/request.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    headers: headers
+  });
+}
+
+export const drugs = formContents => {
+  return post('/api/pml', new FormData(formContents), formHeaders);
+};
+
+export const uris = labels => {
+  return post('/api/uris', JSON.stringify({labels: labels}), defaultHeaders);
```
Comment (houli) [inline panacea/web/static/js/request.js]: {labels}

File: panacea/web/static/js/request.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+};
+
+export const ddis = uris => {
+  return post('/api/ddis', JSON.stringify({drugs: uris}), defaultHeaders);
```
Comment (houli) [inline panacea/web/static/js/request.js]: if we change uris to drugs this could be {drugs}

File: panacea/web/static/js/app.js [hunk 8]
```diff
@@ -1,157 +1,72 @@
...
-      if (labels.length > 0) {
-        displayDrugs(labels);
-
-        const urisResponse = await fetch('/api/uris', {
-          method: 'POST',
-          body: JSON.stringify({labels: labels}),
-          credentials: 'same-origin',
-          headers: defaultHeaders
-        });
+    const drugsResponse = await Request.drugs(this);
```
Comment (c-brenn) [inline panacea/web/static/js/app.js]: any use in keeping this assignment?

Thoughts on this instead
```javascript
handleDrugsResponse(await Request.drugs(this));
```

File: panacea/web/static/js/app.js [hunk 9]
```diff
@@ -1,157 +1,72 @@
...
-
-const displayUnidentifiedDrugs = drugs => {
-  const unidentifiedDrugsTextElement = document.getElementById('unidentified-drugs-text');
-  unidentifiedDrugsTextElement.innerHTML = JSON.stringify(drugs);
-
-  showElement(unidentifiedDrugsPanel);
-};
+async function handleResponse(response, handle) {
+  if (response.ok) {
+    const payload = await response.json();
```
Comment (c-brenn) [inline panacea/web/static/js/app.js]: same thing here

File: panacea/web/static/js/app.js [hunk 10]
```diff
@@ -1,157 +1,72 @@
...
-  const ddisTextElement = document.getElementById('ddis-text');
-  ddisTextElement.innerHTML = JSON.stringify(ddis);
+const handleDrugsResponse = drugsResponse => {
+  handleResponse(drugsResponse, async function ({drugs}) {
+    const labels = drugs.map(x => x.label);
 
-  showElement(ddisPanel);
+    if (labels.length > 0) {
+      View.displayDrugs(labels);
+      const urisResponse = await Request.uris(labels);
```
Comment (c-brenn) [inline panacea/web/static/js/app.js]: and here

File: panacea/web/static/js/app.js [hunk 11]
```diff
@@ -1,157 +1,72 @@
...
-  showElement(errorPanel);
-};
+    if (unidentifiedDrugs.length > 0) {
+      View.displayUnidentifiedDrugs(unidentifiedDrugs);
+    }
 
-const formElement = document.getElementById('file-form');
-const loadingElement = document.getElementById('loading-container');
+    if (uris.length >= 2) {
+      const ddisResponse = await Request.ddis(uris);
```
Comment (c-brenn) [inline panacea/web/static/js/app.js]: and here

**Review by c-brenn — CHANGES_REQUESTED** [review]
nice work 💯

File: panacea/web/static/js/request.js (hunk 7 above)
Comment (22a) [reply panacea/web/static/js/request.js]: they are drug uris as opposed to drug names though, thoughts @c-brenn ?

File: panacea/web/static/js/app.js (hunk 8 above)
Comment (22a) [reply panacea/web/static/js/app.js]: I like this idea, thoughts @houli ?

File: panacea/web/templates/page/index.html.eex [hunk 12]
```diff
@@ -1,21 +1,21 @@
-<%= form_for @conn, pml_path(@conn, :upload), [as: :upload, multipart: true, id: "file-form", class: "center fades-in"], fn f -> %>
+<form accept-charset="UTF-8" action="/api/noop" class="center fades-in" enctype="multipart/form-data" id="file-form" method="post">
   <div class="form-group col-sm-8">
     <div class="input-group">
       <label class="input-group-btn">
         <span class="btn btn-success">
           <span class="glyphicon glyphicon-folder-open" aria-hidden="true">&ensp;</span>
           Choose PML File
-          <%= file_input f, :file, class: "form-control hidden", id: "file-input", required: true %>
+          <input class="form-control hidden" id="file-input" name="upload[file]" required="required" type="file">
```
Comment (houli) [inline panacea/web/templates/page/index.html.eex]: no need for `required="required"` you can just use `required`

**Review by houli — APPROVED** [review]

**Comment by houli** [comment]
🚢 the 🐳

**Review by c-brenn — APPROVED** [review]

**Pull request merged**